### PR TITLE
[FLINK-33208] Support the writable metadata timestamp

### PR DESCRIPTION
**Pull request by Tan-JiaLiang**
## What is the purpose of the change

Currently, the hbase sink does not support write data with `timestamp`, which may cause the data to be written out of order. I suggest to support the timestamp writable metadata for hbase connector so that we can set the `timestamp` when we writing.

## Brief change log

  - *support the writable metadata timestamp*

## Verifying this change

This change is already covered by existing tests, such as
  - *org.apache.flink.connector.hbase1.HBaseConnectorITCase#testTableSinkWithTimestampMetadata*
  - *org.apache.flink.connector.hbase2.HBaseConnectorITCase#testTableSinkWithTimestampMetadata*

### CODE REVIEW

**Comment by MartijnVisser** [comment]
@ferenc-csaky can you also help with this one? I'll also ping you on some of the other HBase PRs 😅

**Comment by Tan-JiaLiang** [comment]
@MartijnVisser Thanks for all the reviewer request, @ferenc-csaky you can review when you are free, I'll always keep an eye on all my PRs until it finish, thank you master!

**Comment by MartijnVisser** [comment]
CC @ferenc-csaky

**Review by ferenc-csaky — COMMENTED** [review]
I think handling `WritableMetadata` as an `enum` makes interactions with the class kinf of hard to read.

I also wondered about is it make sense to handle interactions with `WritableMetadata` as it has multiple different metadata, while we only have `TIMESTAMP` for now? I would introduce those changes when there are other metadata type(s), cause if we do not add any, we have unnecessary complexity.

I opened a PR to your fork where I created a `TimestampMetadata` class and simplified interactions, did some refactor to eliminate duplications.

PTAL

**Comment by Tan-JiaLiang** [comment]
> I think handling WritableMetadata as an enum makes interactions with the class kinf of hard to read.

I didn't think much of it at first, I just followed the WritableMetadata design of [flink-connector-kafka.](https://github.com/apache/flink-connector-kafka). Because I think it's best to be consistent with these kinds of implementations. But I rethink about it, for now the connectors have split into two different projects, and your refactor is better IMO, so +1 for your suggestion.

> I also wondered about is it make sense to handle interactions with WritableMetadata as it has multiple different metadata, while we only have TIMESTAMP for now?

Yes, there also FLINK-30460 should also be supported IMO. After this PR merged, I'll support it ASAP.

I cherry-pick your PR and do a little bit of optimization, PTAL.

And one more question, i want to keep your changes, but i want to squash all the commit in one too (maybe it can good for maintainer merge my PR), any suggestion? Seems git can not keep one commit two authors.

**Comment by ferenc-csaky** [comment]
> And one more question, i want to keep your changes, but i want to squash all the commit in one too (maybe it can good for maintainer merge my PR), any suggestion? Seems git can not keep one commit two authors.

Feel free to squash the commit, if you put a co author line to the end of the commit msg that is perfect:
```
Co-authored-by: Ferenc Csaky <ferenc.csaky@pm.me>
```

**Review by ferenc-csaky — APPROVED** [review]
Thank you, LGTM!

**Comment by Tan-JiaLiang** [comment]
> > And one more question, i want to keep your changes, but i want to squash all the commit in one too (maybe it can good for maintainer merge my PR), any suggestion? Seems git can not keep one commit two authors.
> 
> Feel free to squash the commit, if you put a co author line to the end of the commit msg that is perfect:
> 
> ```
> Co-authored-by: Ferenc Csaky <ferenc.csaky@pm.me>
> ```

Nice! I got it.

**Review by MartijnVisser — APPROVED** [review]

**Pull request merged**
